### PR TITLE
W13: Initial lexer

### PR DESCRIPTION
**Pull request by C0DK**
_**check sub pull requests*'_

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: code/.idea/misc.xml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (C0DK) [inline code/.idea/misc.xml]: REMOVE FILE

File: code/src/main/kotlin/logger/MissingArgumentError.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package logger
+
```
Comment (C0DK) [inline code/src/main/kotlin/logger/MissingArgumentError.kt]: move this file to exceptions/

File: code/src/main/kotlin/logger/TypeError.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * Class used to log type errors.
+ * @param operator The function in which error occurred
+ * @param operands A list of types that are non-compatible
+ */
+class TypeError(lineNumber: Int, lineIndex: Int, lineText: String,
+                private val operator: String, private val operands: List<String>)
+                : ParserException(lineNumber, lineIndex, lineText) {
+    private val types: String = findTypes()
+
+    override val errorType = "TYPE-ERROR"
```
Comment (C0DK) [inline code/src/main/kotlin/logger/TypeError.kt]: add issue for making this automatic (based on class name)

File: code/src/test/kotlin/loggerTests/LoggerTest.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                             " | Function '<' not defined for types: 'num', 'txt' and 'bool'.\n" +
+                             " | -->help: Try casting your types to match eachother.\n", logger.buffer.toString())
+    }
+
+    @org.junit.jupiter.api.Test
+    fun testUndeclaredError() {
+        val error = UndeclaredError(12, 8, "txt s = textString", "textString")
+        val logger = TestLogger()
+        logger.logCompilationError(error)
+        assertEquals("- ERROR: UNDECLARED-ERROR found at line 12 index 8:\n" +
```
Comment (C0DK) [inline code/src/test/kotlin/loggerTests/LoggerTest.kt]: make indents the same. move the first string (in each of these) to the next line.

File: report/2.Analysis/ProblemStatement.tex [hunk 5]
```diff
@@ -0,0 +1,11 @@
+\section{Problem statement}
```
Comment (C0DK) [inline report/2.Analysis/ProblemStatement.tex]: make issue to fix problem statement (based on what INGO said the last time)

File: report/2.Analysis/Scope.tex [hunk 6]
```diff
@@ -0,0 +1,25 @@
+% This section is meant to specify the scope of project, both in terms of the report, the language that will be developed and the compiler.
+% It also includes the Problem Statement
+\section{Scope}
+As stated in the above analysis, the purpose of this project is to design, create and implement a programming language for the Arduino platform that is simple for beginners to use, whilst still teaching the fundamentals of programming.
+The language is meant to enable beginners to quickly and easily write
+code that can be executed, without losing the functionality of a high-level language.
```
Comment (C0DK) [inline report/2.Analysis/Scope.tex]: why is line 5 and 6 two different lines?

File: report/3.Theory/LanguageSpecification.tex [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+"Java" isAwesome // False
+\end{lstlisting}
+
+The main reasoning for including function overloading is simplicity for the user.
+For instance for printing. 
+The print function should be able to print all basic types, so that the user does not have to worry about the type of what he is about to print.
+
+
+\textbf{Generics}\\
+The language will include simple generics into the language.
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: either "include ... in the langauge"
or "introduce ... into the language". currently sounds weird

File: report/3.Theory/LanguageSpecification.tex [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+This is due to some of the fundamental high order functions, like the map function.
+Recall that the map function is a function that takes a list with elements of an arbitrary type along with a modifier function and returns the mapped value of each element in the list.
+In essence this means given an element of type T1 have a function that maps the element to an element of type T2.
+
+In theory this problem could be solved with overloading. 
+However, this would require overloading of a lot of functions with all the possible types within HCL. 
+Since tuples are distinguished by the types within it, there are a theoretical infinite amount of types within HCL, which in turn makes overloading for this case impossible.
+
+To get a better grasp of how generics are used in HCL, consider the following implementation of the map function:
+\begin{lstlisting}
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: use correct listing (just java for now)

File: report/3.Theory/SyntaxTheory.tex [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+First the proper CFG recognizing HCL will be presented, which will be used to proof that HCL is a CFL. 
+Afterwards the most specific computational model for describing HCL will be identified and presented. 
+Lastly the token categories will be described in a computational model. 
+The conclusions of this section will provide the bases for implementing and testing the parser of the HCL compiler. 
+The majority of this section will be based on the EBNF details touched upon in the previous chapter.
+
+\subsection{HCL as an CFL}
+This subsection will investigate whether HCL is a CFL. 
+To do this it is necessary to construct a PDA.
+The reason being that the set of context free languages is equivalent to the set of languages recognized by push down automata.
```
Comment (Abildgaard96) [inline report/3.Theory/SyntaxTheory.tex]: har skrevet den kommentar her før...
når vi har abbreviations for PDA og CFL, skal de bruges <({[HVER GANG]})>

File: report/3.Theory/LexerImplementation.tex [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+It is implemented as a Sealed class. 
+A sealed class in Kotlin, simply specifies that a value can have one specific type, from a predefined set\cite{KotlinSealed}.
+It is essentially an extended enum.
+\begin{lstlisting}[language=java,label=lis:tokenClass,caption=A snippet from the token class .,firstnumber=9]
+sealed class Literal : Token() {
+	class Text(val value: String) : Literal() {
+	override fun toString() = super.toString() + "[$value]"
+	}
+	class Number(strValue: String, val value: Double = strValue.toDouble()) : Literal() {
+	override fun toString() = super.toString() + "[$value]"
```
Comment (C0DK) [inline report/3.Theory/LexerImplementation.tex]: escape the "$" i think (and one more time below)

File: report/3.Theory/LexerImplementation.tex [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+\end{lstlisting}
+
+The bulk of the logic regarding the lexer is done in the lexer class.
+The lexer class implements the ILexer interface.
+
+\begin{lstlisting}[language=java,label=lis:Lexer,caption=The Lexer .,firstnumber=8]
+class Lexer(private val inputContent: String) : ILexer {
+\end{lstlisting}
+
+The ILexer defines a getTokenSequence function, which will give the token sequence and an inputLine function.
```
Comment (C0DK) [inline report/3.Theory/LexerImplementation.tex]: lacks an oxford comma... right? or something - i read it as the getTokenSequence returned an input line function

File: report/3.Theory/LexerImplementation.tex [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+The lexer is comprised of 3 different classes: the token class, the positional token class and the lexer class.
+
+The token class simply contains the type of the token. 
+It is implemented as a Sealed class. 
+A sealed class in Kotlin, simply specifies that a value can have one specific type, from a predefined set\cite{KotlinSealed}.
+It is essentially an extended enum.
+\begin{lstlisting}[language=java,label=lis:tokenClass,caption=A snippet from the token class .,firstnumber=9]
+sealed class Literal : Token() {
+	class Text(val value: String) : Literal() {
+	override fun toString() = super.toString() + "[$value]"
```
Comment (C0DK) [inline report/3.Theory/LexerImplementation.tex]: escape the "$" i think

File: report/3.Theory/SyntaxTheory.tex [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+The length of the string is $p < |s| < (p3)/2$, since $s = a^nbc^n$, where $n = p/2$ and $0 < |b| \leq p/2$.
+
+This means that the third condition is upheld since $|xy|$ cannot be greater than p.
+The second condition is also upheld since y consist of at least some left parentheses. 
+If the string does contain right parentheses, which it does, y cannot be of the length 0.
+
+The first condition is not upheld since if $i = 0$ the new string cannot contain an equal amount of left and right parentheses because of the same reason as to why the second condition is upheld.
+
+This concludes the proof by contradiction.
+Since HCL does not uphold the required property for it to be a regular language, it is not, and the only automaton capable of recognizing the language is the NDPA described in the previous sections.
```
Comment (technight) [inline report/3.Theory/SyntaxTheory.tex]: Maybe emphasize that we have multiple grammars. The language that is to be recognized by the DFA is not equal to the language recognized by the PDA. (PDA recognizes the token sequence (Context free grammar) the DFA recognizes the lexemes ((in our case somewhat) regular grammar))

File: report/Appendix/0.AppendixCtrl.tex [hunk 14]
```diff
@@ -132,4 +134,55 @@ \section{Aalborg Tekniske Gymnasium Student and Teacher Interview (In danish)}
 
 - Denne gruppe acede det. Stadig en smule problemer med associering.
 
-- Sidste gruppe var rimelig god, men stadig lidt tvivl om associering.
\ No newline at end of file
+- Sidste gruppe var rimelig god, men stadig lidt tvivl om associering.
+
+\section{Extended Backus-Naur Form of HCL grammar}
+\label{AppendixEBNF}
```
Comment (C0DK) [inline report/Appendix/0.AppendixCtrl.tex]: this should be updated to include the current EBNF

File: report/3.Theory/LexerImplementation.tex (hunk 11 above)
Comment (Lambcreek) [reply report/3.Theory/LexerImplementation.tex]: Yes, a comma after "sequence" would improve the sentence a lot

File: report/3.Theory/Lexer.tex [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+\textbf{ANTLR}\\
+ANTLR is a scanner/parser generator, which when given a context-free grammar expressed using Extended Backus-Naur form (EBNF), can generate lexers, parsers and lexer-parsers\cite{ANTLR}.
+ANTLR can output to Java, C\#, Python2, Python3, Javascript, Go, C++ and Swift\cite{ANTLRDocsTargets}.
+One of the primary advantages of ANTLR is that it uses a single notation for both lexers and parsers\cite{ANTLRDocs}.
+This means that after the lexer has been implemented, it would be extremely easy to implement the parser afterwards, as it would be as simple as specifying that a parser needs to be generated.
+As one of the primary design decisions of ANTLR is "simplicity over complexity", as well as the fact that ANTLR is often used in academia, it would be a good choice for a lexer/parser generator, as it can be assumed that it is easy to use.
+ANTLR has official plugins for both Intellij, NetBeans and Eclipse, which improves the user experience, by supporting syntax highlighting, formatting and error detection.
+
+\textbf{Lex}\\
+At the 4th semester, the \textit{Crafting a compiler} uses Lex as an example of a Lexer generator\cite{CraftingACompiler}.
```
[the remark this answers is not in the document]
Comment (C0DK) [reply report/3.Theory/Lexer.tex]: the "textit" pretty much does that

[the remark this answers is not in the document]
Comment (Lambcreek) [reply report/3.Theory/Lexer.tex]: But saying "The _Crafting a compiler_" doesn't make sense.
Either remove "the" or specify that it is a textbook.

File: report/3.Theory/SyntaxTheory.tex (hunk 9 above)
Comment (jones1994) [reply report/3.Theory/SyntaxTheory.tex]: Logikken var:
CFL -> Context free language
ikke: context free languages
Men hvis vi kan skrive CFL's og det er  fint så gør vi bare det

Det er blevet fikset i et commit så er alt er godt nu :)

**Pull request merged**